### PR TITLE
Add option to enable locking debug logging

### PR DESCRIPTION
**Pull request by icewind1991**
Logs all attempts to lock, unlock or change logs on files.

### CODE REVIEW

**Review by MorrisJobke — COMMENTED** [review]
Tested and works 👍

**Review by MorrisJobke — APPROVED** [review]

**Review by LukasReschke — APPROVED** [review]
Looks good to me and works.

@icewind1991 I adjusted this to also include the storage ID in the log entry. Any objections?

**Comment by LukasReschke** [comment]
Also the tests claim to fail now?

```
1) Test\Files\ViewTest::testReadFromWriteLockedPath with data set #0 ('/testuser/{folder}', '')
1019s
110
Failed asserting that exception of type "\OCP\Lock\LockedException" is thrown.
1019s
112
2) Test\Files\ViewTest::testReadFromWriteLockedPath with data set #1 ('/testuser', '/{folder}')
1019s
113
Failed asserting that exception of type "\OCP\Lock\LockedException" is thrown.
1019s
115
3) Test\Files\ViewTest::testReadFromWriteLockedPath with data set #2 ('', '/testuser/{folder}')
1019s
116
Failed asserting that exception of type "\OCP\Lock\LockedException" is thrown.
1019s
118
4) Test\Files\ViewTest::testWriteToReadLockedFile with data set #0 ('/testuser/{folder}', '')
1019s
119
Failed asserting that exception of type "\OCP\Lock\LockedException" is thrown.
1019s
121
5) Test\Files\ViewTest::testWriteToReadLockedFile with data set #1 ('/testuser', '/{folder}')
1019s
122
Failed asserting that exception of type "\OCP\Lock\LockedException" is thrown.
1019s
124
6) Test\Files\ViewTest::testWriteToReadLockedFile with data set #2 ('', '/testuser/{folder}')
1019s
125
Failed asserting that exception of type "\OCP\Lock\LockedException" is thrown.
1019s
127
7) Test\Files\ViewTest::testLockLocalMountPointPathInsteadOfStorageRoot
1019s
128
Mount point path was locked on root storage
1019s
129
Failed asserting that false is true.
1019s
131
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1358
1019s
133
8) Test\Files\ViewTest::testLockStorageRootButNotLocalMountPoint
1019s
134
Mount point storage root was locked on original storage
1019s
135
Failed asserting that false is true.
1019s
137
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1385
1019s
139
9) Test\Files\ViewTest::testLockMountPointPathFailReleasesBoth
1019s
140
Cannot acquire shared lock because storage root is already locked
1019s
141
Failed asserting that false is true.
1019s
143
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1415
1019s
145
10) Test\Files\ViewTest::testChangeLock
1019s
146
Failed asserting that false is true.
1019s
148
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1477
1019s
150
11) Test\Files\ViewTest::testLockBasicOperation with data set #0 ('touch', array('touch-create.txt'), 'touch-create.txt', 'create', 1, 2, 1)
1019s
151
File locked properly during pre-hook
1019s
152
Failed asserting that null matches expected 1.
1019s
154
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1800
1019s
156
12) Test\Files\ViewTest::testLockBasicOperation with data set #1 ('fopen', array('test-write.txt', 'w'), 'test-write.txt', 'write', 1, 2, null, 2)
1019s
157
File locked properly during pre-hook
1019s
158
Failed asserting that null matches expected 1.
1019s
160
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1800
1019s
162
13) Test\Files\ViewTest::testLockBasicOperation with data set #2 ('mkdir', array('newdir'), 'newdir', 'write', 1, 2, 1)
1019s
163
File locked properly during pre-hook
1019s
164
Failed asserting that null matches expected 1.
1019s
166
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1800
1019s
168
14) Test\Files\ViewTest::testLockBasicOperation with data set #3 ('file_put_contents', array('file_put_contents.txt', 'blah'), 'file_put_contents.txt', 'write', 1, 2, 1)
1019s
169
File locked properly during pre-hook
1019s
170
Failed asserting that null matches expected 1.
1019s
172
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1800
1019s
174
15) Test\Files\ViewTest::testLockBasicOperation with data set #4 ('rmdir', array('dir'), 'dir', 'delete', 1, 2, 1)
1019s
175
File locked properly during pre-hook
1019s
176
Failed asserting that null matches expected 1.
1019s
178
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1800
1019s
180
16) Test\Files\ViewTest::testLockBasicOperation with data set #5 ('unlink', array('test.txt'), 'test.txt', 'delete', 1, 2, 1)
1019s
181
File locked properly during pre-hook
1019s
182
Failed asserting that null matches expected 1.
1019s
184
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1800
1019s
186
17) Test\Files\ViewTest::testLockBasicOperation with data set #6 ('file_get_contents', array('test.txt'), 'test.txt', 'read', 1, 1, null)
1019s
187
File locked properly during pre-hook
1019s
188
Failed asserting that null matches expected 1.
1019s
190
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1800
1019s
192
18) Test\Files\ViewTest::testLockBasicOperation with data set #7 ('fopen', array('test.txt', 'r'), 'test.txt', 'read', 1, 1, null)
1019s
193
File locked properly during pre-hook
1019s
194
Failed asserting that null matches expected 1.
1019s
196
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1800
1019s
198
19) Test\Files\ViewTest::testLockBasicOperation with data set #8 ('opendir', array('dir'), 'dir', 'read', 1, 1, null)
1019s
199
File locked properly during pre-hook
1019s
200
Failed asserting that null matches expected 1.
1019s
202
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1800
1019s
204
20) Test\Files\ViewTest::testLockFilePutContentWithStream
1019s
205
File locked properly during pre-hook
1019s
206
Failed asserting that null matches expected 1.
1019s
208
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1842
1019s
210
21) Test\Files\ViewTest::testLockFopen
1019s
211
File locked properly during pre-hook
1019s
212
Failed asserting that null matches expected 1.
1019s
214
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:1880
1019s
216
22) Test\Files\ViewTest::testLockFileRename with data set #0 ('rename', 2)
1019s
217
Source file locked properly during pre-hook
1019s
218
Failed asserting that null matches expected 1.
1019s
220
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:2029
1019s
222
23) Test\Files\ViewTest::testLockFileRename with data set #1 ('copy', 1)
1019s
223
Source file locked properly during pre-hook
1019s
224
Failed asserting that null matches expected 1.
1019s
226
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:2029
1019s
228
24) Test\Files\ViewTest::testLockFileRenameUnlockOnException
1019s
229
Target file is locked before operation
1019s
230
Failed asserting that null matches expected 2.
1019s
232
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:2100
1019s
234
25) Test\Files\ViewTest::testLockFileRenameCrossStorage with data set #0 ('rename', 'moveFromStorage', 2)
1019s
235
Source file locked properly during pre-hook
1019s
236
Failed asserting that null matches expected 1.
1019s
238
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:2208
1019s
240
26) Test\Files\ViewTest::testLockFileRenameCrossStorage with data set #1 ('copy', 'copyFromStorage', 1)
1019s
241
Source file locked properly during pre-hook
1019s
242
Failed asserting that null matches expected 1.
1019s
244
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:2208
1019s
246
27) Test\Files\ViewTest::testLockMoveMountPoint
1019s
247
Source path locked properly during pre-hook
1019s
248
Failed asserting that null matches expected 1.
1019s
250
/drone/src/github.com/nextcloud/server/tests/lib/Files/ViewTest.php:2264
1019s
252
28) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\UploadTest::testUploadOverWriteReadLocked
1019s
253
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
255
29) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\UploadTest::testUploadOverWriteWriteLocked
1019s
256
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
258
30) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\UploadTest::testChunkedUploadOutOfOrderReadLocked
1019s
259
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
261
31) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\UploadTest::testChunkedUploadOutOfOrderWriteLocked
1019s
262
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
264
32) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\PartFileInRootUploadTest::testUploadOverWriteReadLocked
1019s
265
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
267
33) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\PartFileInRootUploadTest::testUploadOverWriteWriteLocked
1019s
268
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
270
34) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\PartFileInRootUploadTest::testChunkedUploadOutOfOrderReadLocked
1019s
271
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
273
35) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\PartFileInRootUploadTest::testChunkedUploadOutOfOrderWriteLocked
1019s
274
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
276
36) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\EncryptionUploadTest::testUploadOverWriteReadLocked
1019s
277
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
279
37) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\EncryptionUploadTest::testUploadOverWriteWriteLocked
1019s
280
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
282
38) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\EncryptionUploadTest::testChunkedUploadOutOfOrderReadLocked
1019s
283
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
285
39) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\EncryptionUploadTest::testChunkedUploadOutOfOrderWriteLocked
1019s
286
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
288
40) OCA\DAV\Tests\unit\Connector\Sabre\RequestTest\DownloadTest::testDownloadWriteLocked
1019s
289
Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
1019s
291
41) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testSimplePutFailsMoveFromStorage
1019s
292
Failed asserting that false is true.
1019s
294
/drone/src/github.com/nextcloud/server/apps/dav/tests/unit/Connector/Sabre/FileTest.php:606
1019s
296
42) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testChunkedPutFailsFinalRename
1019s
297
Failed asserting that false is true.
1019s
299
/drone/src/github.com/nextcloud/server/apps/dav/tests/unit/Connector/Sabre/FileTest.php:644
1019s
301
43) OCA\DAV\Tests\unit\Connector\Sabre\FileTest::testPutLocking
1019s
302
File was locked during pre-hooks
1019s
303
Failed asserting that false is true.
1019s
305
/drone/src/github.com/nextcloud/server/apps/dav/tests/unit/Connector/Sabre/FileTest.php:924
1019s
307
44) OCA\Files_Sharing\Tests\LockingTest::testLockAsRecipient
1019s
308
Failed asserting that exception of type "\OCP\Lock\LockedException" is thrown.
1019s
310

```

**Review by MorrisJobke — APPROVED** [review]

File: lib/private/Files/Storage/Common.php [hunk 1]
```diff
@@ -681,7 +685,28 @@ public function getMetaData($path) {
...
+					'app' => 'locking',
+				]
+			);
+		}
+		try {
+			$provider->acquireLock('files/' . md5($this->getId() . '::' . trim($path, '/')), $type);
+		} catch (LockedException $e) {
+			if ($logger) {
+				$logger->logException($e);
+			}
```
Comment (LukasReschke) [inline lib/private/Files/Storage/Common.php]: Throw missing here?

**Review by LukasReschke — CHANGES_REQUESTED** [review]

File: lib/private/Files/Storage/Common.php (hunk 1 above)
Comment (MorrisJobke) [inline lib/private/Files/Storage/Common.php]: Right 🙈

File: lib/private/Files/Storage/Common.php [hunk 2]
```diff
@@ -699,7 +745,36 @@ public function releaseLock($path, $type, ILockingProvider $provider) {
...
+					$this->getId()
+				),
+				[
+					'app' => 'locking',
+				]
+			);
+		}
+		try {
+			$provider->changeLock('files/' . md5($this->getId() . '::' . trim($path, '/')), $type);
+		} catch (LockedException $e) {
```
Comment (LukasReschke) [inline lib/private/Files/Storage/Common.php]: Shouldn't the PHPDoc here and at the other places then note that this may throw an exception?

**Comment by MorrisJobke** [comment]
@LukasReschke Should be fixed now.

**Review by LukasReschke — APPROVED** [review]

File: lib/private/Files/Storage/Common.php [hunk 3]
```diff
@@ -681,25 +685,101 @@ public function getMetaData($path) {
 	 * @throws \OCP\Lock\LockedException
 	 */
 	public function acquireLock($path, $type, ILockingProvider $provider) {
-		$provider->acquireLock('files/' . md5($this->getId() . '::' . trim($path, '/')), $type);
+		$logger = $this->getLockLogger();
+		if ($logger) {
+			$typeString = ($type === ILockingProvider::LOCK_SHARED) ? 'shared' : 'exclusive';
+			$logger->info(
+				sprintf(
```
Comment (nickvergessen) [inline lib/private/Files/Storage/Common.php]: logging it self does sprintf btw, so you can do:
```php
$logger->info(
	'acquire {type} lock on "{path}" on storage "{storage}"',
	[
		'type' => $typeString,
		'path' => $path,
		'storage' => $this->getId()
		'app' => 'locking',
	]
);
```

JFYI

**Pull request merged**